src/json.ts

```diff
@@ -0,0 +1,355 @@
+}
+
+// If byte data is parseable to a valid unicode string then do so
+// otherwise parse the byte data to a hex string
+export function bytesToJSONValue(value: Bytes): JSONValue {
+  // fallback - assume the data is a hex string (always valid)
+  let result = json.try_fromString('["' + value.toHexString() + '"]');
+  // If the bytes can be parsed as a string, then losslessly re-encoded into
+  // UTF-8 bytes, then consider a valid UTF-8 encoded string and store
+  // string value in json.
+  // note: Bytes.toString() uses WTF-8 encoding as opposed to UTF-8.  Solidity
+  // encodes UTF-8 strings, so safe assume any string data are UTF-8 encoded.
+  let stringValue: string = value.toString();
+  let reEncodedBytesValue: Bytes = Bytes.fromUTF8(stringValue);
+  if (reEncodedBytesValue.toHexString() == value.toHexString()) {
+    // if the bytes are the same then the string was valid UTF-8
+    let potentialResult = json.try_fromString('["' + stringValue + '"]');
+    if (potentialResult.isOk) {
+      result = potentialResult;
+    }
+  }
+  return result.value.toArray()[0];
+}
+
+/*** Utils for creating and updating TypedMap<string, JSONValue> objects ***/
+export function createUpdatedTypedMapWithEntryAdded<ValueType>(
+  object: TypedMap<string, JSONValue>,
+  key: string,
+  value: ValueType
+): TypedMap<string, JSONValue> {
+  const updated = json.fromString(typedMapToJSONString(object)).toObject();
+
+  let jsonValue: JSONValue = toJSONValue(value);
+
+  updated.set(key, jsonValue);
+  return updated;
+}
+
+export function createUpdatedTypedMapWithEntryRemoved(
+  object: TypedMap<string, JSONValue>,
+  key: string
+): TypedMap<string, JSONValue> {
+  const withRemovedTypedMap: TypedMap<string, JSONValue> = new TypedMap();
+
+  for (let i = 0; i < object.entries.length; i++) {
+    let entry = object.entries[i];
+    if (entry.key != key) {
+      withRemovedTypedMap.set(entry.key.toString(), entry.value);
+    }
+  }
+
+  return withRemovedTypedMap;
+}
+
+/**
+ * @dev note this function will overwrite the value at the given key if
+ * it already exists and is not an array.
+ */
+export function createUpdatedTypedMapWithArrayValueAdded<ValueType>(
+  object: TypedMap<string, JSONValue>,
+  key: string,
+  value: ValueType
+): TypedMap<string, JSONValue> {
+  const stringified = typedMapToJSONString(object);
+  const updated = json.fromString(stringified).toObject();
+
+  let currentValue = object.get(key);
+
+  // Convert json value to array so we can add the new value
+  let arr =
+    currentValue && currentValue.kind === JSONValueKind.ARRAY
+      ? currentValue.toArray()
+      : [];
+
+  let modifiedArray = createModifiedArray(arr, value, ArrayOperation.ADD);
+  let newValue = jsonValueArrayToJSONValue(modifiedArray);
+
+  updated.set(key, newValue);
+
+  return updated;
+}
+
+/**
+ * @dev note this function will overwrite the value at the given key if
+ * it already exists and is not an array.
+ */
```

Comment:
This will actually overwrite an existing non-array value with an empty array. Not sure if we want to keep that behavior for symmetry with the add function or not?

Reply:
Ah, good call out. Symmetry with the add function makes sense to me.